Dynatrace/dynatrace-operator
# Use 1 validator per CR

## Description

[K8S-11317](https://dt-rnd.atlassian.net/browse/K8S-11317)

We initially wanted to have the validation logic per version, which makes sense on paper, but not very realistic in reality, as most new validation logic needs to be added to all versions, and sometimes have to think about cross-version checks.

So now instead we use 1 custom-validator per CR, that can handle multiple version.
- For older versions, just use `ConvertTo`

We still need the 1 validation endpoint per version in the `ValidationWebhookConfiguration`
- It's how the framework works:
   - You define the webhook for a given version of the CR, so when the webhook receives the request it will convert it to that version.
     - So if you define only 1 webhook and point all the versions there, all but the 1 version will succeed, others will just fail due to a conversion error within the framework.

Sidenote: Will break https://github.com/Dynatrace/dynatrace-operator/pull/3853 but I can live with that, should be an easy merge conflict. (it's also my PR :P )

## How can this be tested?

Deploy 1 dk/ec for each version (broken or not) and check if the webhook reacts accordingly

Example: 3 different dk versions, all intentionally broken, and the webhook rejects them accordingly
```
Warning: ActiveGate specification missing memory limits. Can cause excess memory usage.
Error from server (Forbidden): error when creating "./.vscode/deploy/dynakube_zib50933.yaml": admission webhook "v1beta3.dynakube.webhook.dynatrace.com" denied the request:
1 error(s) found in the DynaKube
 1. The length limit for the name of a DynaKube is 40, because it is the base for the name of resources related to the DynaKube. (example: dkName-activegate-<some-hash>)
	The limit is necessary because kubernetes uses the name of some resources (example: StatefulSet) for the label value, which has a limit of 63 characters. (see https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)
Error from server (Forbidden): error when creating "./.vscode/deploy/dynakube_zib50933.yaml": admission webhook "v1beta2.dynakube.webhook.dynatrace.com" denied the request:
1 error(s) found in the DynaKube
 1. The DynaKube's specification has an invalid API URL value set.
	Make sure you correctly specify the URL in your custom resource (including the /api postfix).

Error from server (Forbidden): error when creating "./.vscode/deploy/dynakube_zib50933.yaml": admission webhook "v1beta1.dynakube.webhook.dynatrace.com" denied the request:
1 error(s) found in the DynaKube
 1. The DynaKube's specification has an invalid API URL value set.
	Make sure you correctly specify the URL in your custom resource (including the /api postfix).
```
